### PR TITLE
fix issue #79211

### LINKED ISSUE #79211: daemonset status not correct  num of status some filed
**What happened**:
in mysql cluster, it's has 33 node and every node running a pod of kube-proxy created by daemonset and all the pod of kube-proxy ds was running status. but when l describe the ds of kube-proxy, the status like this:
```
currentNumberScheduled: 4 
numberMisscheduled: 29 
desiredNumberScheduled: 4 
numberReady: 4  
updatedNumberScheduled: 4  
numberAvailable: 4  
```
what 's maening of the ds status filed？ l can‘t  find some doc or source code to know it. 
so l think those of myself like this: 
```
currentNumberScheduled: current daemoncontroller simulate schedule paas  and the pod not on the node in crrent time. (but when the controller manger restart, it's my less or zero)

numberMisscheduled: current daemoncontroller simulate schedule not paas. 

desiredNumberScheduled：only current daemoncontroller simulate schedule paas.

numberReady:  the pod status condition ready is true

updatedNumberScheduled:  current daemoncontroller simulate schedule paas  and the pod not on the node in crrent time.  the pod status condition available is true,the pod status condition ready is true , and the pod hash was update

numberAvailable:  the pod status condition available is true

```
please notic the all num was base on the current status of damon-controller.  when ths controller was down and restart, some num was not correct

the code of the status update owner by the daemonsetController like this:

```
func (dsc *DaemonSetsController) updateDaemonSetStatus(ds *apps.DaemonSet, hash string, updateObservedGen bool) error {
	klog.V(4).Infof("Updating daemon set status")
	nodeToDaemonPods, err := dsc.getNodesToDaemonPods(ds)
	if err != nil {
		return fmt.Errorf("couldn't get node to daemon pod mapping for daemon set %q: %v", ds.Name, err)
	}

	nodeList, err := dsc.nodeLister.List(labels.Everything())
	if err != nil {
		return fmt.Errorf("couldn't get list of nodes when updating daemon set %#v: %v", ds, err)
	}

	var desiredNumberScheduled, currentNumberScheduled, numberMisscheduled, numberReady, updatedNumberScheduled, numberAvailable int
	for _, node := range nodeList {
		wantToRun, _, _, err := dsc.nodeShouldRunDaemonPod(node, ds)
		if err != nil {
			return err
		}

		scheduled := len(nodeToDaemonPods[node.Name]) > 0

		if wantToRun {
			desiredNumberScheduled++
			if scheduled {
				currentNumberScheduled++
				// Sort the daemon pods by creation time, so that the oldest is first.
				daemonPods, _ := nodeToDaemonPods[node.Name]
				sort.Sort(podByCreationTimestampAndPhase(daemonPods))
				pod := daemonPods[0]
				if podutil.IsPodReady(pod) {
					numberReady++
					if podutil.IsPodAvailable(pod, ds.Spec.MinReadySeconds, metav1.Now()) {
						numberAvailable++
					}
				}
				// If the returned error is not nil we have a parse error.
				// The controller handles this via the hash.
				generation, err := util.GetTemplateGeneration(ds)
				if err != nil {
					generation = nil
				}
				if util.IsPodUpdated(pod, hash, generation) {
					updatedNumberScheduled++
				}
			}
		} else {
			if scheduled {
				numberMisscheduled++
			}
		}
	}
	numberUnavailable := desiredNumberScheduled - numberAvailable

	err = storeDaemonSetStatus(dsc.kubeClient.AppsV1().DaemonSets(ds.Namespace), ds, desiredNumberScheduled, currentNumberScheduled, numberMisscheduled, numberReady, updatedNumberScheduled, numberAvailable, numberUnavailable, updateObservedGen)
	if err != nil {
		return fmt.Errorf("error storing status for daemon set %#v: %v", ds, err)
	}

	return nil
}
```

1. first to simulate schedule the ds's pod to the node
2. second get the pod of  ds on the current node, if pod exist scheduled is true
3. if current schedule paas but the pod oready exist on the node, check the ready and avalible
4. step 3 not paas, but step 2 pass, numberMisscheduled++

l think the logic of the check ready and available is error.
 
**What you expected to happen**:
maybe the  currentNumberScheduled,numberMisscheduled,desiredNumberScheduled,updatedNumberScheduled  can do like this, but the numberReady,numberAvailable shouldn't do like this.

l think should consider the pod which exsit on the node.  the code like this:

```
func (dsc *DaemonSetsController) updateDaemonSetStatus(ds *apps.DaemonSet, hash string, updateObservedGen bool) error {
	klog.V(4).Infof("Updating daemon set status")
	nodeToDaemonPods, err := dsc.getNodesToDaemonPods(ds)
	if err != nil {
		return fmt.Errorf("couldn't get node to daemon pod mapping for daemon set %q: %v", ds.Name, err)
	}

	nodeList, err := dsc.nodeLister.List(labels.Everything())
	if err != nil {
		return fmt.Errorf("couldn't get list of nodes when updating daemon set %#v: %v", ds, err)
	}

	var desiredNumberScheduled, currentNumberScheduled, numberMisscheduled, numberReady, updatedNumberScheduled, numberAvailable int
	for _, node := range nodeList {
		wantToRun, _, _, err := dsc.nodeShouldRunDaemonPod(node, ds)
		if err != nil {
			return err
		}

		scheduled := len(nodeToDaemonPods[node.Name]) > 0

		if wantToRun {
			desiredNumberScheduled++
			if scheduled {
				currentNumberScheduled++

			}
		} else {
			if scheduled {
				numberMisscheduled++
			}
		}
		
		// Sort the daemon pods by creation time, so that the oldest is first.
		daemonPods, _ := nodeToDaemonPods[node.Name]
		sort.Sort(podByCreationTimestampAndPhase(daemonPods))
		pod := daemonPods[0]
		if podutil.IsPodReady(pod) {
			numberReady++
			if podutil.IsPodAvailable(pod, ds.Spec.MinReadySeconds, metav1.Now()) {
				numberAvailable++
			}
		}
		// If the returned error is not nil we have a parse error.
		// The controller handles this via the hash.
		generation, err := util.GetTemplateGeneration(ds)
		if err != nil {
			generation = nil
		}
		if util.IsPodUpdated(pod, hash, generation) {
			updatedNumberScheduled++
		}
	}
	numberUnavailable := desiredNumberScheduled - numberAvailable

	err = storeDaemonSetStatus(dsc.kubeClient.AppsV1().DaemonSets(ds.Namespace), ds, desiredNumberScheduled, currentNumberScheduled, numberMisscheduled, numberReady, updatedNumberScheduled, numberAvailable, numberUnavailable, updateObservedGen)
	if err != nil {
		return fmt.Errorf("error storing status for daemon set %#v: %v", ds, err)
	}

	return nil
}
```
**How to reproduce it (as minimally and precisely as possible)**:
create a ds in the cluster,  when the ds's all pod running, stop the controller-manager. 
**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 
1.14.2
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
- Kernel (e.g. `uname -a`):
Linux k8s-master 3.10.0-862.3.3.el7.x86_64 #1 SMP Fri Jun 15 04:15:27 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
baniry
- Network plugin and version (if this is a network-related bug):
flannel:v0.11.0
- Others:

### PR DESCRIPTION
**Pull request by huangjiasingle**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
> /kind bug

**What this PR does / why we need it**:
fix issue #79211
**Which issue(s) this PR fixes**:

Fixes #79211

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by huangjiasingle** [comment]
@mortent can review the logic of the code?

File: pkg/controller/daemon/daemon_controller.go [hunk 1]
```diff
@@ -1175,8 +1175,11 @@ func (dsc *DaemonSetsController) updateDaemonSetStatus(ds *apps.DaemonSet, nodeL
 
 		if wantToRun {
 			desiredNumberScheduled++
-			if scheduled {
+			if !scheduled {
```
Comment (mortent) [inline pkg/controller/daemon/daemon_controller.go]: I don't follow the logic here. If a pod belonging to a daemonset should be scheduled on a node and it is, then it seems like the `currentNumberScheduled` should be incremented.

File: pkg/controller/daemon/daemon_controller.go [hunk 2]
```diff
@@ -1200,6 +1203,27 @@ func (dsc *DaemonSetsController) updateDaemonSetStatus(ds *apps.DaemonSet, nodeL
 		} else {
 			if scheduled {
 				numberMisscheduled++
+				desiredNumberScheduled++
```
Comment (mortent) [inline pkg/controller/daemon/daemon_controller.go]: This section doesn't seem right. If pod is scheduled onto a node where it should not be, I think we should only count that one in the `numberMisscheduled` variable. In other words, if the pod is misscheduled, we no longer cares if the pod is ready or available.

**Review by mortent — COMMENTED** [review]
I don't quite follow why this change is needed. Looking at the original bug, the status for the daemonset is what I would expect to see. I'm not sure why so many pods are misscheduled in the example, but I wouldn't expect misscheduled pods to be counted as ready and/or available.

File: pkg/controller/daemon/daemon_controller.go (hunk 1 above)
Comment (huangjiasingle) [reply pkg/controller/daemon/daemon_controller.go]: you may think the ds controller running forever.  currentNumberScheduled  = wantToRun but not exist on every node.  do you agree with this logic?

**Pull request closed without merging**